vslib/src/SwitchStateBaseMACsec.cpp

```diff
@@ -0,0 +1,859 @@
+        SWSS_LOG_ERROR("Attribute is empty");
+        return SAI_STATUS_FAILURE;
+    }
+
+    const sai_attribute_t *attr = nullptr;
+    SAI_METADATA_GET_ATTR_BY_ID(attr, SAI_MACSEC_PORT_ATTR_MACSEC_DIRECTION, attr_count, attr_list);
+    macsec_attr.m_direction = attr->value.s32;
+
+    SAI_METADATA_GET_ATTR_BY_ID(attr, SAI_MACSEC_PORT_ATTR_PORT_ID, attr_count, attr_list);
+    auto port_id = attr->value.oid;
+    CHECK_STATUS(findHostInterfaceInfoByPort(port_id, macsec_attr.m_info));
+    macsec_attr.m_veth_name = vs_get_veth_name(macsec_attr.m_info->m_name, port_id);
+    macsec_attr.m_macsec_name = SAI_VS_MACSEC_PREFIX + macsec_attr.m_veth_name;
+
+    return SAI_STATUS_SUCCESS;
+}
+
+sai_status_t SwitchStateBase::loadMACsecAttrFromMACsecSC(
+    _In_ sai_object_id_t object_id,
+    _In_ uint32_t attr_count,
+    _In_ const sai_attribute_t *attr_list,
+    _Out_ MACsecAttr &macsec_attr)
+{
+    SWSS_LOG_ENTER();
+
+    if (attr_list == nullptr || attr_count == 0)
+    {
+        SWSS_LOG_ERROR("Attribute is empty");
+        return SAI_STATUS_FAILURE;
+    }
+
+    const sai_attribute_t *attr = nullptr;
+
+    SAI_METADATA_GET_ATTR_BY_ID(attr, SAI_MACSEC_SC_ATTR_MACSEC_DIRECTION, attr_count, attr_list);
+    macsec_attr.m_direction = attr->value.s32;
+
+    SAI_METADATA_GET_ATTR_BY_ID(attr, SAI_MACSEC_SC_ATTR_MACSEC_SCI, attr_count, attr_list);
+    auto sci = attr->value.u64;
+    std::stringstream sci_convert;
+    sci_convert << std::hex << bswap_64(sci);
+    macsec_attr.m_sci = sci_convert.str();
+
+    SAI_METADATA_GET_ATTR_BY_ID(attr, SAI_MACSEC_SC_ATTR_MACSEC_EXPLICIT_SCI_ENABLE, attr_count, attr_list);
+    macsec_attr.m_send_sci = attr->value.booldata;
+
+    SAI_METADATA_GET_ATTR_BY_ID(attr, SAI_MACSEC_SC_ATTR_FLOW_ID, attr_count, attr_list);
+    auto flow_id = attr->value.oid;
+
+    sai_object_id_t port_id = SAI_NULL_OBJECT_ID;
+    CHECK_STATUS(findPortByMACsecFlow(flow_id, port_id));
+    CHECK_STATUS(findHostInterfaceInfoByPort(port_id, macsec_attr.m_info));
+    macsec_attr.m_veth_name = vs_get_veth_name(macsec_attr.m_info->m_name, port_id);
+    macsec_attr.m_macsec_name = SAI_VS_MACSEC_PREFIX + macsec_attr.m_veth_name;
+
+    return SAI_STATUS_SUCCESS;
+}
+
+sai_status_t SwitchStateBase::loadMACsecAttrFromMACsecSA(
+    _In_ sai_object_id_t object_id,
+    _In_ uint32_t attr_count,
+    _In_ const sai_attribute_t *attr_list,
+    _Out_ MACsecAttr &macsec_attr)
+{
+    SWSS_LOG_ENTER();
+
+    if (attr_list == nullptr || attr_count == 0)
+    {
+        SWSS_LOG_ERROR("Attribute is empty");
+        return SAI_STATUS_FAILURE;
+    }
+
+    const sai_attribute_t *attr = nullptr;
+    std::vector<sai_attribute_t> attrs;
+
+    SAI_METADATA_GET_ATTR_BY_ID(attr, SAI_MACSEC_SA_ATTR_SC_ID, attr_count, attr_list);
+
+    // Find SCI and MACsec flow
+    attrs.clear();
+    attrs.emplace_back();
+    attrs.back().id = SAI_MACSEC_SC_ATTR_FLOW_ID;
+    attrs.emplace_back();
+    attrs.back().id = SAI_MACSEC_SC_ATTR_MACSEC_SCI;
```

Comment:
std::vector<sai_attribute_t> attrs(2);

attrs[0].id = SAI_MACSEC_SC_ATTR_FLOW_ID;
attrs[1].id = SAI_MACSEC_SC_ATTR_MACSEC_SCI;

3 lines instead of 6, clear is not needed, back will get iterator when you can access directly by index